packages/file-search/src/node/file-search-service-impl.ts

```diff
@@ -0,0 +1,106 @@
+        const basePath = FileUri.fsPath(new URI(uri));
+        const opts = {
+            fuzzyMatch: true,
+            limit: Number.MAX_SAFE_INTEGER,
+            useGitignore: true,
+            ...options
+        };
+        const result: string[] = [];
+        const limitReached = new Error("limit reached");
+        const filtered = this.getFiltered(basePath, opts);
+        try {
+            this.findRecursive(basePath, filtered, filePath => {
+                if (result.length >= opts.limit) {
+                    throw limitReached;
+                }
+                if (opts.fuzzyMatch && fuzzy.test(searchPattern, filePath)) {
+                    result.push(FileUri.create(filePath).toString());
+                } else {
+                    if (filePath.toLocaleLowerCase().indexOf(searchPattern.toLocaleLowerCase()) !== -1) {
+                        result.push(FileUri.create(filePath).toString());
+                    }
+                }
+            });
+        } catch (e) {
+            if (e !== limitReached) {
+                throw e;
+            }
+        }
+        return result;
+    }
+
+    private findRecursive(filePath: string, filtered: Filtered, acceptor: (fileName: string) => void) {
+        const result = fs.readdirSync(filePath);
+        for (const child of result) {
+            const childPath = path.join(filePath, child);
+            if (!filtered(child, childPath)) {
+                if (fs.statSync(childPath).isDirectory()) {
+                    this.findRecursive(childPath, filtered, acceptor);
+                } else {
+                    acceptor(childPath);
+                }
+            }
+        }
+    }
+
+    private getFiltered(basePath: string, options: { fuzzyMatch: boolean, limit: number, useGitignore: boolean }): Filtered {
+        const defaultFilter = (simpleName: string) =>
+            simpleName === '.git' ||
+            simpleName.startsWith('.git/');
+        if (options.useGitignore) {
+            const gitIgnore = this.findGitIgnore(basePath);
+            if (gitIgnore) {
+                const matcher = parser.compile(gitIgnore.contents);
+                return (short, long) => defaultFilter(short) || matcher.denies(short);
+            }
+        }
+        return defaultFilter;
+    }
+
+    private findGitIgnore(basePath: string): { path: string, contents: string } | undefined {
+        try {
+            const fullPath = path.join(basePath, '.gitignore');
+            const result = fs.readFileSync(fullPath, 'utf-8');
+            if (this.logger.isInfo()) {
+                this.logger.info(`Found gitignore below ${fullPath}`);
+            }
+            return {
+                path: basePath,
+                contents: result
+            };
+        } catch {
+            try {
+                const parent = path.resolve(basePath, '..');
```

Comment:
I'm not sure it's wise to look outside the WS root.
The git extension doesn't do that either. And I think it may lead to weird situations.

Reply:
It's the right thing to do. Just as git does.
What weird situations?